datastore/datastore.go

```diff
@@ -0,0 +1,260 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	hook "github.com/alanshaw/ipfs-hookds"
+	hopts "github.com/alanshaw/ipfs-hookds/opts"
+	hres "github.com/alanshaw/ipfs-hookds/query/results"
+	hropts "github.com/alanshaw/ipfs-hookds/query/results/opts"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	levelds "github.com/ipfs/go-ds-leveldb"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/routing"
+	"github.com/libp2p/go-libp2p-kad-dht/providers"
+	"github.com/libp2p/hydra-booster/metrics"
+	"github.com/multiformats/go-base32"
+	"github.com/whyrusleeping/timecache"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+)
+
+// root namespace of provider keys
+var providersRoot = datastore.NewKey(providers.ProvidersKeyPrefix)
+
+// AddProviderFunc adds a provider for a given CID to the datastore
+type AddProviderFunc func(ctx context.Context, c cid.Cid, id peer.ID)
+
+// GetRoutingFunc is a function that returns an appropriate routing module given a CID
+type GetRoutingFunc func(cid.Cid) (routing.Routing, AddProviderFunc, error)
```

Comment:
What is the origin of this interface?

Reply:
We have to create the datastore before the DHT, so passing a function that retrieves the DHT module allows us to create the datastore before the DHT and access it later.